contracts/src/TeleporterTokenSource.sol

```diff
@@ -134,30 +192,67 @@ abstract contract TeleporterTokenSource is ITeleporterTokenBridge, TeleporterOwn
         bytes32 sourceBlockchainID,
         address originSenderAddress,
         bytes memory message
-    ) internal virtual override {
-        (SendTokensInput memory input, uint256 amount) =
-            abi.decode(message, (SendTokensInput, uint256));
+    ) internal override {
+        BridgeMessage memory bridgeMessage = abi.decode(message, (BridgeMessage));
 
         // Check that bridge instance returning has sufficient amount in balance
         uint256 senderBalance = bridgedBalances[sourceBlockchainID][originSenderAddress];
-        require(senderBalance >= amount, "TeleporterTokenSource: insufficient bridge balance");
+        require(
+            senderBalance >= bridgeMessage.amount,
+            "TeleporterTokenSource: insufficient bridge balance"
+        );
 
         // Decrement the bridge balance by the unwrap amount
-        bridgedBalances[sourceBlockchainID][originSenderAddress] = senderBalance - amount;
-
-        // If the final destination of the transfer is this source bridge instance, withdraw the tokens.abi
-        // Otherwise, perform a multihop transfer to the input destination bridge instance.
-        if (input.destinationBlockchainID == blockchainID) {
-            require(
-                input.destinationBridgeAddress == address(this),
-                "TeleporterTokenSource: invalid bridge address"
+        bridgedBalances[sourceBlockchainID][originSenderAddress] =
+            senderBalance - bridgeMessage.amount;
+
+        if (bridgeMessage.messageType == BridgeMessageType.SINGLE_HOP_SEND) {
+            SingleHopSendMessage memory payload =
+                abi.decode(bridgeMessage.payload, (SingleHopSendMessage));
+            emit TokensWithdrawn(payload.recipient, bridgeMessage.amount);
+            _withdraw(payload.recipient, bridgeMessage.amount);
+            return;
+        } else if (bridgeMessage.messageType == BridgeMessageType.SINGLE_HOP_CALL) {
+            SingleHopCallMessage memory payload =
+                abi.decode(bridgeMessage.payload, (SingleHopCallMessage));
+            _handleSendAndCall(payload, bridgeMessage.amount);
```

Comment:
I'm not sure I follow the suggestion, but it may just be due to background context. `_sendAndCall` is called to issue an `SendAndCall` message on the chain the transfer is being initiated, and `_handleSendAndCall` is called on the chain receiving the transfer to handle the execution of the message that is being received.